### PR TITLE
Disables non existent projects on the recent projects list

### LINKED ISSUE #28655: Deactivate "Open ""... entry in "Recent projects" list for unavailable projects
Author Name: **Harrissou Santanna** (@DelazJ)
Original Redmine Issue: [20836](https://issues.qgis.org/issues/20836)
Affected QGIS version: 3.4.2
Redmine category:gui

---
When opening a new session of QGIS, The "Recent projects" list will show unavailable/lost project in a grayed color. A right-click over these files allows to refresh (refresh what?)  or remove the file from the list.
But there's also an *Open ""...* entry (should be "open directory..." if file was available) that does nothing. It could be nice to deactivate/hide that entry if the file is not accessible and give it a sexy label.

---

- [unavailableProject.png](https://issues.qgis.org/attachments/download/13988/unavailableProject.png) (Harrissou Santanna)

### LINKED ISSUE #31011: "Project-Open Recent" does not detect if a *.qgz file is missing (but does for *.qgs)
Save a project as *.qgs and *.qgz.
Rename the two project files in your file manager.
Open another project that isn't empty.
Try to open the *.qgs project file from "Project-Open Recent".  You will get a suitable error message and your window will redraw the project that that you were in.  OK, good.
Try to open the *.qgz project file from "Project-Open Recent" - it will think it has succeeded, creating a new empty project and updating the QGIS window title to match the name of the missing *.qgs
If you click on "Project-properties" you will see the project path matches the missing *.qgz, and if you do "Project-save" it will create the missing *.qgz.

### LINKED ISSUE #31385: Open project folder and select project file does not work for non file system based projects
**Describe the bug**

We have a handy button to open project's folder.

open project folder

This button does not work for Postgresql stored projects. For GeoPackage stored projects, it does makes sense (open the geopackage's folder).

The  three different project syntax options are:

```C
    /* Project name syntax:
     *
     * /home/qgis/projects/ortos_cim_2019.qgz
     * geopackage:/home/qgis/projects/natural.gpkg?projectName=natural%20earth
     * postgresql:?service=pg_geotuga&sslmode=disable&dbname=&schema=ortos&project=teste
     */
```

**How to Reproduce**

For GeoPackage and Postgresql stored projects, just click on the open project button.
There is an inoffensive error reported by the file system manager.

**QGIS and OS versions**

QGIS 3.8

**Additional context**

I'm already preparing a PR to address this issue and issue #28655

### PR DESCRIPTION
**Pull request by jgrocha**
## Description

This PR improves the code to be more aware of the project syntax. The project is usually a path, but it can also be a provider stored project.

Project name syntax:

```C
     /*
     * /home/qgis/projects/ortos_cim_2019.qgz
     * geopackage:/home/qgis/projects/natural.gpkg?projectName=natural%20earth
     * postgresql:?service=pg_geotuga&sslmode=disable&dbname=&schema=ortos&project=teste
     */
```

In this PR we use `QgsApplication::projectStorageRegistry()->projectStorageFromUri( )` to get the correspondent storage class.

To check if the project still exist, we check if the file is still there, for file system stored projects.

For storage based projects:
- **postgresql** - We don't check if they still exist. You may have authentication problems, can be disconnected right now, etc.
- **geopackage** We check if the file still exist. In fact, we don't have to test it, because this is already done in the provider code.

Fix #31011
Fix #31385
Fix #28655

## Checklist

- [X] Commit messages are descriptive and explain the rationale for changes
- [X] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [X] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [X] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Comment by nirvn** [comment]
Lots of duplicate code. I think this could benefit from being a static function, either in qgsproject or a new qgsprojectutils class, which would return a qvariantmap object in which a filepath item is not empty if the project is file based.

QVariantMap QgsProject::decodeUri( const QString & )

We're doing that to decode providers' uri and it's proven very useful.

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -4271,7 +4271,25 @@ void QgisApp::updateRecentProjectPaths()
                               : QFileInfo( recentProject.path ).completeBaseName(), QDir::toNativeSeparators( recentProject.path )
                             ).replace( "&", "&&" )
                       );
-    //action->setEnabled( QFile::exists( ( recentProject.path ) ) );
+
+    QRegularExpression re_gpkg( "^(geopackage:)([^\?]+)\?(.+)$", QRegularExpression::CaseInsensitiveOption );
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp]: Use camelCase for C++ code

```suggestion
    QRegularExpression reGpkg( "^(geopackage:)([^\?]+)\?(.+)$", QRegularExpression::CaseInsensitiveOption );
```

**Comment by jgrocha** [comment]
I've review the changes and now I'm taking advantage of QgsProjectStorage class.

### Recent project menu entries

In the recent projects menu, we disable file system based entries that does not exist and also geopackage stored projects, when the geopackage does not exist.

I've added a small warning icon.

missing recent projects menu

### Recent project entries in the welcome page

The same for the welcome page.

missing projects welcome page

### Project properties

The button to open the project folder is also improved. It does nothing for Postgresql stored projects.

For GeoPackage stored projects and file system stored projects, the forlder in open in the file manager and the file is selected.

**Comment by nirvn** [comment]
Much nicer.

**Comment by jgrocha** [comment]
> Much nicer.

Thanks for the push @nirvn. Your comment challenged me to to it better.

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -4271,7 +4271,27 @@ void QgisApp::updateRecentProjectPaths()
                               : QFileInfo( recentProject.path ).completeBaseName(), QDir::toNativeSeparators( recentProject.path )
                             ).replace( "&", "&&" )
                       );
-    //action->setEnabled( QFile::exists( ( recentProject.path ) ) );
+
+    QgsProjectStorage *storage = QgsApplication::projectStorageRegistry()->projectStorageFromUri( recentProject.path );
+
+    if ( storage )
+    {
+      QString path = storage->filePath( recentProject.path );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Should this also be resetting recentProject.path to match?

Comment (jgrocha) [reply src/app/qgisapp.cpp]: Can you elaborate more @nyalldawson? I didn't understood your point.

Comment (nyalldawson) [reply src/app/qgisapp.cpp]: Actually sorry, I had the wrong place for this comment. It actual relates to QgsWelcomePage::showContextMenuForProjects, where the project path is used to open the file in explorer. We'd need similar handling there to use the GPKG file path...

**Comment by nirvn** [comment]
@jgrocha , do you think you can push this through the finish line prior to 3.10 release?

**Comment by jgrocha** [comment]
> @jgrocha , do you think you can push this through the finish line prior to 3.10 release?

Yes, I will try to finish it today.

**Comment by nirvn** [comment]
Great.

File: src/app/qgisapp.cpp (hunk 2 above)
Comment (jgrocha) [reply src/app/qgisapp.cpp]: Thanks @nyalldawson 
Now the context menu is aware of the project type (last commit https://github.com/qgis/QGIS/pull/31386/commits/17b73bfe7628d950cdda9fc2f563ffb5da993dda)
If the geopackage file is missing, users can open the closed path (I got it almost for free from the existing code).

**Comment by jgrocha** [comment]
The context menu is now aware of the project storage type.

For geopackage stored projects, the geopackage folder is opened and the geopackage file is selected, as shown in the following screen shot.

show folder and select geopackage

For non existing geopackage files (the file is missing from the original location), the closest path is shown. 

Note for myself: The closest path does not behave as expected if the path contains spaces, for example, as shown in the print screen. Existent folder are discarded, even if they exist. This can be fixed later.

show geopackage closest folder

Finally, postgresql storage based projects do not have the contextual Open directory... available, as expected.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**